sdks/python/apache_beam/tools/utils.py

```diff
@@ -34,3 +41,87 @@ def check_compiled(module):
         "Profiling uncompiled code.\n"
         "To compile beam, run "
         "'pip install Cython; python setup.py build_ext --inplace'")
+
+
+class BenchmarkConfig(
+    collections.namedtuple(
+        "BenchmarkConfig", ["benchmark", "size", "num_runs"])):
+  """
+  Attributes:
+    benchmark: a callable that takes an int argument - benchmark size,
+      and returns a callable. A returned callable must run the code being
+      benchmarked on an input of specified size.
+
+      For example, one can implement a benchmark as:
+
+      class MyBenchmark(object):
+        def __init__(self, size):
+          [do necessary initialization]
+        def __call__(self):
+          [run the code in question]
+
+    size: int, a size of the input. Aggregated per-element metrics
+      are counted based on the size of the input.
+    num_runs: int, number of times to run each benchmark.
+  """
+
+  def __str__(self):
+    return "%s, %s element(s)" % (
+        getattr(self.benchmark, '__name__', str(self.benchmark)),
+        str(self.size))
+
+
+def run_benchmarks(benchmark_suite, verbose=True):
+  """Runs benchmarks, and collects execution times.
+
+  A simple instrumentation to run a callable several times, collect and print
+  its execution times.
+
+  Args:
+    benchmark_suite: A list of BenchmarkConfig.
+    verbose: bool, whether to print benchmark results to stdout.
+
+  Returns:
+    A dictionary of the form string -> list of floats. Keys of the dictionary
+    are benchmark names, values are execution times in seconds for each run.
+  """
+
+  def run(benchmark_fn, size):
+    # Contain each run of a benchmark inside a function so that any temporary
+    # objects can be garbage-collected after the run.
+    benchmark_instance_callable = benchmark_fn(size)
+    start = time.time()
+    _ = benchmark_instance_callable()
+    return time.time() - start
+
+  cost_series = collections.defaultdict(list)
+  for benchmark_config in benchmark_suite:
+    name = str(benchmark_config)
+    num_runs = benchmark_config.num_runs
+    size = benchmark_config.size
+    for run_id in range(num_runs):
+      # Do a proactive GC before each run to minimize side-effects of different
+      # runs.
+      gc.collect()
+      time_cost = run(benchmark_config.benchmark, size)
+      cost_series[name].append(time_cost)
+      if verbose:
+        per_element_cost = time_cost / size
+        print("%s: run %d of %d, per element time cost: %g sec" % (
+            name, run_id + 1, num_runs, per_element_cost))
+    if verbose:
+      print("")
+
+  if verbose:
+    pad_length = max([len(str(bc)) for bc in benchmark_suite])
+
+    for benchmark_config in benchmark_suite:
+      name = str(benchmark_config)
+      per_element_median_cost = (
+          numpy.median(cost_series[name]) / benchmark_config.size)
+      std = numpy.std(cost_series[name]) / benchmark_config.size
+
+      print("%s: per element median time cost: %g sec, std: %g sec" % (
```

Comment:
Scale the std by the median (or give it as a percentage) so it's easier to see if the variance is significant.

Reply:
Done.
Sample output:
```
small_int, FastPrimitivesCoder, 1000 element(s)                                   : per element median time cost: 7.60555e-08 sec, relative std: 27.47%
large_int, FastPrimitivesCoder, 1000 element(s)                                   : per element median time cost: 1.50919e-07 sec, relative std: 3.13%
small_string, FastPrimitivesCoder, 1000 element(s)                                : per element median time cost: 4.40478e-07 sec, relative std: 2.62%
large_string, FastPrimitivesCoder, 1000 element(s)                                : per element median time cost: 8.64029e-07 sec, relative std: 3.34%
small_list, FastPrimitivesCoder, 1000 element(s)                                  : per element median time cost: 7.69496e-07 sec, relative std: 3.20%
small_list, IterableCoder[FastPrimitivesCoder], 1000 element(s)                   : per element median time cost: 7.24435e-07 sec, relative std: 3.05%
large_list, FastPrimitivesCoder, 1000 element(s)                                  : per element median time cost: 5.04484e-05 sec, relative std: 1.27%
large_list, IterableCoder[FastPrimitivesCoder], 1000 element(s)                   : per element median time cost: 5.01723e-05 sec, relative std: 0.92%
small_tuple, FastPrimitivesCoder, 1000 element(s)                                 : per element median time cost: 2.95877e-07 sec, relative std: 3.43%
large_tuple, FastPrimitivesCoder, 1000 element(s)                                 : per element median time cost: 5.34335e-05 sec, relative std: 1.22%
small_dict, FastPrimitivesCoder, 1000 element(s)                                  : per element median time cost: 1.57547e-06 sec, relative std: 5.36%
large_dict, FastPrimitivesCoder, 1000 element(s)                                  : per element median time cost: 1.9261e-05 sec, relative std: 8.33%
wv_with_one_window, WindowedValueCoder[FastPrimitivesCoder], 1000 element(s)      : per element median time cost: 2.5202e-05 sec, relative std: 0.77%
wv_with_multiple_windows, WindowedValueCoder[FastPrimitivesCoder], 1000 element(s): per element median time cost: 0.000845308 sec, relative std: 11.42%
```